src/extension/provider.ts

```diff
@@ -0,0 +1,126 @@
+import { parse } from "diff2html";
+import { basename } from "path";
+import * as vscode from "vscode";
+import { MessageToExtensionHandler, MessageToWebview } from "../shared/message";
+import { APP_CONFIG_SECTION, extractConfig } from "./configuration";
+import { MessageToExtensionHandlerImpl } from "./message/handler";
+import { buildSkeleton, SkeletonElementIds } from "./skeleton";
+
+interface DiffViewerProviderArgs {
+  extensionContext: vscode.ExtensionContext;
+  webviewPath: string;
+}
+
+interface WebviewContext {
+  document: vscode.TextDocument;
+  panel: vscode.WebviewPanel;
+}
+
+export class DiffViewerProvider implements vscode.CustomTextEditorProvider {
+  private static readonly VIEW_TYPE = "diffViewer";
+  private messageReceivedHandler: MessageToExtensionHandler | undefined;
+
+  public constructor(private readonly args: DiffViewerProviderArgs) {}
+
+  public static register(args: DiffViewerProviderArgs): vscode.Disposable {
+    return vscode.window.registerCustomEditorProvider(DiffViewerProvider.VIEW_TYPE, new DiffViewerProvider(args), {
+      webviewOptions: {
+        retainContextWhenHidden: true,
+      },
+      supportsMultipleEditorsPerDocument: false,
+    });
+  }
+
+  public async resolveCustomTextEditor(
+    diffDocument: vscode.TextDocument,
+    webviewPanel: vscode.WebviewPanel,
+    token: vscode.CancellationToken
+  ): Promise<void> {
+    if (token.isCancellationRequested) {
+      return;
+    }
+
+    const webviewUri = webviewPanel.webview.asWebviewUri(
+      vscode.Uri.joinPath(this.args.extensionContext.extensionUri, this.args.webviewPath)
+    );
+
+    webviewPanel.webview.options = {
+      enableScripts: true,
+    };
+
+    webviewPanel.webview.html = buildSkeleton(webviewUri);
+
+    this.messageReceivedHandler = new MessageToExtensionHandlerImpl({
```

Comment:
If I understand it correctly, `MessageToExtensionHandlerImpl` is specific to a `diffDocument`, but the extension has only one instance so storing it in `this` means we only have the message handler for the last opened document (the one created the last time `resolveCustomTextEditor` was called), correct?

Your refactor has made the code more complex than before (not meant as a criticism), so I'm getting into it slowly and I'm now stumbling on the fact that a click in one diff document's `webView` can send a message to a `MessageToExtensionHandlerImpl` with a different `diffDocument`.

Please correct me if I just missed something @caponetto 8-)

Reply:
Just merged https://github.com/caponetto/vscode-diff-viewer/pull/51. Hopefully, this will be enough. :)